src/rgw/rgw_lc_tier.cc

```diff
@@ -0,0 +1,1242 @@
+          /* continue with upload anyway */
+        }
+        ldout(tier_ctx.cct, 0) << "sync of object=" << tier_ctx.obj << " via multipart upload, finished sending part #" << status.cur_part << " etag=" << status.parts[status.cur_part].etag << dendl;
+      }
+
+      yield call(new RGWLCCompleteMultipartCR(tier_ctx.cct, tier_ctx.http_manager, tier_ctx.conn.get(), dest_obj, status.upload_id, status.parts));
+      if (retcode < 0) {
+        ldout(tier_ctx.cct, 0) << "ERROR: failed to complete multipart upload of obj=" << tier_ctx.obj << " (error: " << cpp_strerror(-retcode) << ")" << dendl;
+        ret_err = retcode;
+        yield call(new RGWLCStreamAbortMultipartUploadCR(tier_ctx, dest_obj, status_obj, status.upload_id));
+        return set_cr_error(ret_err);
+      }
+
+      /* remove status obj */
+      ret_err = tier_ctx.store->delete_system_obj(status_obj.pool, status_obj.oid, nullptr, null_yield);
+
+      if (ret_err < 0) {
+        ldout(tier_ctx.cct, 0) << "ERROR: failed to abort multipart upload obj=" << tier_ctx.obj << " upload_id=" << status.upload_id << " part number " << status.cur_part << " (" << cpp_strerror(-ret_err) << ")" << dendl;
+        /* ignore error, best effort */
+      }
+      return set_cr_done();
+    }
+    return 0;
+  }
+};
+
+int RGWLCCloudCheckCR::operate() {
+  /* Check if object has already been transitioned */
+  rgw_lc_obj_properties obj_properties(tier_ctx.o.meta.mtime, tier_ctx.o.meta.etag,
+                                       tier_ctx.o.versioned_epoch, tier_ctx.acl_mappings,
+                                       tier_ctx.target_storage_class);
+
+  RGWBucketInfo b;
+  string target_obj_name;
+  int reterr = 0;
+
+  b.bucket.name = tier_ctx.target_bucket_name;
+  target_obj_name = tier_ctx.bucket_info.bucket.name + "/" +
+                    (*tier_ctx.obj)->get_name();
+    if (!tier_ctx.o.is_current()) {
+      target_obj_name += get_key_instance((*tier_ctx.obj)->get_key());
+    }
+
+  std::unique_ptr<rgw::sal::RGWBucket> dest_bucket;
+  std::unique_ptr<rgw::sal::RGWObject> dest_obj;
+
+
+  reterr = tier_ctx.store->get_bucket(nullptr, b, &dest_bucket);
+  if (reterr < 0) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_bucket - " << tier_ctx.target_bucket_name << " , reterr = " << reterr << dendl;
+    return reterr;
+  }
+  
+  dest_obj = dest_bucket->get_object(rgw_obj_key(target_obj_name));
+  if (!dest_obj) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_object path - " << target_obj_name << dendl;
+    return -1;
+  }
+
+  std::shared_ptr<RGWLCStreamGetCRF> get_crf;
+  get_crf.reset(new RGWLCStreamGetCRF((CephContext *)(tier_ctx.cct), get_env(), this,
+        (RGWHTTPManager*)(tier_ctx.http_manager),
+        obj_properties, tier_ctx.conn, static_cast<rgw::sal::RGWObject *>(dest_obj.get())));
+  int ret = 0;
+
+  reenter(this) {
```

Comment:
note that every time this `RGWLCCloudCheckCR` coroutine yields, it will resume with a new call to `operate()`. we use the `reenter` macro, so that we can resume from where we left off

by placing all of this code outside of the `reenter` block, that means it gets re-executed every time we resume. it looks like all of this should move inside, so it only runs once at the beginning?

reenter is implemented with a switch statement, which means that local variables can be an issue; generally you need to make them member variables of the coroutine class instead

Reply:
done..

Reply:
I applied these changes to other coroutines as well and that fixed couple of issues I was facing with yield. Thanks! Updated the PR with latest code. Please review.